common/nwg_classes.h

```diff
@@ -87,3 +101,168 @@ struct DesktopEntry {
+    void fullscreen();
+    template <typename S> void show(S);
+private:
+    std::variant<LayerShell, SwayShell, GenericShell> shell;
+};
+
+template <typename Hint>
+void GenericShell::show(CommonWindow& window, Hint hint) {
+    window.show();
+    window.set_type_hint(Gdk::WINDOW_TYPE_HINT_SPLASHSCREEN);
+    window.set_decorated(false);
+    auto [d_x, d_y, d_w, d_h] = geometry(window);
+    auto window_coord_at_side = [](auto d_size, auto w_size, auto side, auto margin) {
+        return !side * margin + side * (d_size - w_size - margin);
+    };
+    Overloaded place_window {
+        [](hint::Auto_) { /* we assume the window is opened at center... */ },
+        [&](hint::Fullscreen_) { window.fullscreen(); },
+        [&,d_w=d_w,d_h=d_h](hint::Side<hint::Horizontal> hint) {
+            int w_x = 0, w_y = 0;
+            window.get_position(w_x, w_y);
+            w_x = window_coord_at_side(d_w, window.get_width(), hint.side, hint.margin);
+            window.move(w_x, (d_h - window.get_height()) / 2);
+        },
+        [&,d_w=d_w,d_h=d_h](hint::Side<hint::Vertical> hint) {
+            int w_x = 0, w_y = 0;
+            window.get_position(w_x, w_y);
+            w_y = window_coord_at_side(d_h, window.get_height(), hint.side, hint.margin);
+            window.move((d_w - window.get_width()) / 2, w_y);
+        },
+        [&,d_w=d_w,d_h=d_h](hint::Sides hint) {
+            auto w_x = window_coord_at_side(d_w, window.get_width(), hint.h.side, hint.h.margin);
+            auto w_y = window_coord_at_side(d_h, window.get_height(), hint.v.side, hint.v.margin);
+            window.move(w_x, w_y);
+        }
+    };
+    place_window(hint);
+    window.present();   // grab focus
+}
+
+#ifdef HAVE_GTK_LAYER_SHELL
+template <typename Hint>
+void LayerShell::show(CommonWindow& window, Hint hint) {
+    std::array<bool, 4> edges{ 0, 0, 0, 0 };
+    std::array<int,  4> margins{ 0, 0, 0, 0 };
+    constexpr Overloaded index { [](hint::Horizontal){ return 0; }, [](hint::Vertical) { return 2; } };
+    auto account_side = [&](auto side) {
+        constexpr auto i = index(side.side_type);
+        edges[i + side.side] = true;
+        margins[i + side.side] = side.margin;
+    };
+    Overloaded set_edges_margins {
+        [&](hint::Auto_) { /* nothing to do */ },
+        [&](hint::Fullscreen_) { edges = { 1, 1, 1, 1 }; },
+        [&](hint::Sides hint) { account_side(hint.h); account_side(hint.v); },
+        account_side
+    };
+    set_edges_margins(hint);
+    window.show();
+    auto gtk_win = window.gobj();
+    std::array edges_ {
+        GTK_LAYER_SHELL_EDGE_LEFT,
+        GTK_LAYER_SHELL_EDGE_RIGHT,
+        GTK_LAYER_SHELL_EDGE_TOP,
+        GTK_LAYER_SHELL_EDGE_BOTTOM
+    };
+    for (size_t i = 0; i < 4; i++) {
+        gtk_layer_set_anchor(gtk_win, edges_[i], edges[i]);
+        gtk_layer_set_margin(gtk_win, edges_[i], margins[i]);
+    }
+    gtk_layer_set_layer(gtk_win, GTK_LAYER_SHELL_LAYER_TOP);
```

Comment:
This could be changed to `GTK_LAYER_SHELL_LAYER_OVERLAY` if you want to see launchers on top of fullscreen windows. `TOP` is rendered behind fullscreen windows.

See also https://github.com/lbonn/rofi/issues/12

Reply:
If possible, I'd like the grid to respect bar's/panel's exclusive zone.

Reply:
`gtk_layer_set_exclusive_zone(gtk_win, 0)` would do that.
layer: overlay with exclusive zone: -1 will draw a fullscreen window over any bars or exclusive zones. I'd personally prefer that option.

***
To be honest, bars are not supposed to use top layer. The main reason to do so is when your compositor renders layer shell popup at the same layer as the bar (i.e. hidden under normal windows), and it was decided that this behavior is not correct.
Sway already shows any layer-shell popups above regular windows or top layer surfaces.

Reply:
> overlay with exclusive zone: -1 will draw a fullscreen window over any bars or exclusive zones

Acceptable.

> bars are not supposed to use top layer

But they do, if user tells them to.